File: 2016/Day19/main.cpp
#include <cstdint>
#include <iostream>

uint32_t highest_power_of_2(uint32_t n) {
  uint32_t p = 1;
  while (p <= n / 2)
    p *= 2;
  return p;
}

uint32_t part2(uint32_t num_elves) {
  uint32_t p3 = 1;
  while (p3 <= num_elves / 3)
    p3 *= 3;

  if (num_elves == p3)
    return num_elves;

  if (num_elves <= 2 * p3)
    return num_elves - p3;

  return 2 * num_elves - 3 * p3;
}

int main() {
  const uint32_t num_elves = 3017957;

  uint32_t p = highest_power_of_2(num_elves);

  uint32_t l = num_elves - p;

  uint32_t winner = 2 * l + 1;

  std::cout << "Part 1: " << winner << std::endl;

  std::cout << "Part 2: " << part2(num_elves) << std::endl;

  return 0;
}

File: 2016/Day19/main.lua
---@param num_elves integer
---@return integer
local function part1(num_elves)
  local highest_power_of_2 = 1
  while highest_power_of_2 * 2 <= num_elves do
    highest_power_of_2 = highest_power_of_2 * 2
  end
  local l = num_elves - highest_power_of_2
  return 2 * l + 1
end

local function part2(num_elves)
  local p3 = 1
  while p3 * 3 <= num_elves do
    p3 = p3 * 3
  end

  if num_elves == p3 then
    return num_elves
  elseif num_elves <= 2 * p3 then
    return num_elves - p3
  else
    return 2 * num_elves - 3 * p3
  end
end

local num_elves = 3017957

print('Part 1: ' .. part1(num_elves))
print('Part 2: ' .. part2(num_elves))
